backend/src/controllers/__tests__/CMSTutorialController.unit.test.ts

```diff
@@ -0,0 +1,202 @@
+import Server from "../../App";
+import supertest from "supertest";
+import { MongoClient, Db, Collection } from "mongodb";
+import { BAD_REQUEST_TEXT, BAD_OBJECTID_PARSE_TEXT } from "../../constants";
+import PartialTutorial from "../../dtos/PartialTutorial.dto";
+import Tutorial from "../../dtos/Tutorial.dto";
+
+/**
+ * This test utilises [supertest]{@link https://www.npmjs.com/package/supertest}
+ * and initialises the full server to perform the test per test.
+ *
+ * @author ale8k
+ */
+describe("CMSTutorialController", () => {
+    let connection: MongoClient;
+    const usersCollectionName = "cms-users";
+    const tutorialsCollectionName = "tutorials";
+    let db: Db;
+    let usersCollection: Collection;
+    let tutorialsCollection: Collection;
+
+    beforeAll(async () => {
+        process.env.PORT = "3003";
+        process.env.DB_URL = global.__MONGO_URI__;
+        process.env.DB_NAME = global.__MONGO_DB_NAME__;
+        process.env.BCRYPT_SALT = "12";
+        process.env.JWT_KEY = "iamakeylol";
+        process.env.JWT_EXPIRY = "2 days";
+
+        connection = await MongoClient.connect((global.__MONGO_URI__), {
+            poolSize: 10,
+            useNewUrlParser: true,
+            useUnifiedTopology: true
+        });
+
+        db = await connection.db(global.__MONGO_DB_NAME__);
+        usersCollection = db.collection(usersCollectionName);
+        tutorialsCollection = db.collection(tutorialsCollectionName);
+        await usersCollection.insert(
+            {
+                username: "alex",
+                password: "$2b$12$Cp/IwyOUsKiJZTANWklBB.k4mv07lIV1gSLT4FbtsOI.eoFi2qfTu"
+            }
+        );
+    });
+
+    /**
+     * /CMS/TUTORIALS/CREATE
+     */
+    it("cms/tutorials/create should create a tutorial, with a 200 response", async (done) => {
+        await new Server({ path: ".test.env" }).setupServer().then((app) => {
+            let jwt: string;
+            const agent = supertest(app);
+            // Grab our JWT by logging in
+            agent.post("/cms/auth/login").send({ username: "alex", password: "p123" })
+                // Utilise supertest's callback to pass the JWT
+                .expect(response => {
+                    jwt = response.body.jwt;
+                }).end(() => {
+                    agent.post("/cms/tutorials/create").set("Authorization", `Bearer ${jwt}`)
+                        .send(new PartialTutorial("tutorial", "html", "markdown", "category"))
+                        .expect(201).end(() => {
+                            agent.get("/cms/tutorials/all").set("Authorization", `Bearer ${jwt}`).expect((response) => {
+                                const tutArr = response.body as Tutorial[];
+                                const tut = tutArr[0];
+                                console.log(tut);
+                                expect(tut.isAvailable).toBeTruthy();
+                            }).end(done);
+                        });
+                });
```

Comment:
this is too high level for me. havent seen done twice. No idea about supertest. Explain me that alright?

Reply:
Aha sure yeah, it's performing chained HTTP requests following our app structure, so it logs in, creates one and then tries to get them all. That's all XD